### PR TITLE
Type support API mods

### PR DESCRIPTION
**Pull request by kuznetsovmoci**

### CODE REVIEW

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h [hunk 1]
```diff
@@ -18,31 +18,45 @@
...
+  */
   void * (*create_replier)(
-    void * participant,
-    const char * request_topic_str,
-    const char * response_topic_str,
-    const void * datareader_qos, const void * datawriter_qos,
-    void ** reader, void ** writer, void * (*allocator)(size_t));
-  // Function to destroy a replier
+    DDS::DomainParticipant_var dds_participant,
+    const char* request_topic_str,
```
Comment (mitza-oci) [inline rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: Determine what our "Style Guide" is, and be consistent `const char* s` vs `const char * s` (I don't like the latter since it looks like binary operator).

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h [hunk 2]
```diff
@@ -18,31 +18,45 @@
...
+    allocator_t allocator);
+  //! Function to destroy a requester
   const char * (*destroy_requester)(
-    void * untyped_requester, void (* deallocator)(void *));
-  // Function to create a replier
+    void * untyped_requester, deallocator_t deallocator);
+  //! Function to create a replier
+  /*!
+  Set pub/sub default writer/reader qos to the ros converted qos, before calling this function
+  TODO: add 'const' to parameters, when determined they stay unchanged by the Requestor
```
Comment (mitza-oci) [inline rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: This is probably not needed now that the _var types are here.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h [hunk 3]
```diff
@@ -18,31 +18,45 @@
...
+    DDS::Subscriber_var dds_subscriber,
+    allocator_t allocator);
+  //! Function to destroy a requester
   const char * (*destroy_requester)(
-    void * untyped_requester, void (* deallocator)(void *));
-  // Function to create a replier
+    void * untyped_requester, deallocator_t deallocator);
+  //! Function to create a replier
+  /*!
+  Set pub/sub default writer/reader qos to the ros converted qos, before calling this function
```
Comment (mitza-oci) [inline rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: Instead of talking about "ROS converting" (what might happen before calling this), document the assumptions of the implementation.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 1 above)
Comment (kuznetsovmoci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: I agree with your preference but the rest of code uses other style, so I'll keep its style.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h [hunk 4]
```diff
@@ -18,31 +18,45 @@
 #include <stdint.h>
 #include <rmw/types.h>
 #include "rosidl_generator_c/service_type_support_struct.h"
+#include "dds/DCPS/Service_Participant.h"
+
```
Comment (lij-oci) [inline rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h:22]: In general, avoid including implementation-specific headers in header files reduces unwanted dependencies. That's the reason we decided using opaque types (void*) here for OpenDDS objects to avoid potential problems. If this is properly verified as unnecessary, then I would suggest using references for the OpenDDS objects here.

File: rosidl_typesupport_opendds_cpp/resource/srv__type_support.cpp.em [hunk 5]
```diff
@@ -127,22 +125,20 @@ int64_t send_request__@(service.namespaced_type.name)(
...
+    DDS::DomainParticipant_var dds_participant,
+    const char* request_topic_str,
+    const char* response_topic_str,
+    DDS::Publisher_var dds_publisher,
+    DDS::Subscriber_var dds_subscriber,
+    allocator_t allocator)
 {
 @# TODO: Implement, considering original code in ffe10f9 or earlier
     return NULL;
 }
```
Comment (lij-oci) [inline rosidl_typesupport_opendds_cpp/resource/srv__type_support.cpp.em:137]: on failure, return nullptr; consistently.

Comment (lij-oci) [inline rosidl_typesupport_opendds_cpp/resource/srv__type_support.cpp.em:137]: on failure, return nullptr; consistently.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 4 above)
Comment (mitza-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h:22]: I don't think this is a problem.  We are not depending on any implementation details of rosidl_typesupport here, just the common OpenDDS headers that are available to all.  I will defer to others on ROS2 build / packaging system details.

The one change I would suggest would be to `#include` the specific OpenDDS headers that are needed instead of `Service_Participant.h`.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 1 above)
Comment (mitza-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: As long as we're consistent (especially within the same file), I'm OK with it

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 4 above)
Comment (kuznetsovmoci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h:22]: In earlier discussions I've got an impression that passing OpenDDS types would be preferable, if it's easy. That would save us type casting / conversion. Should I go back and change all types to 'untyped'?

Comment (mitza-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h:22]: No, at least not for now.  If you have a build locally that's compiling all of what should be compiled, that's probably a good sign that it can work.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 3 above)
Comment (kuznetsovmoci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: How does this sound: "Reader / Writer QoS of the passed Publisher / Subscriber must be set to desired values prior to calling this function"?

Comment (mitza-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: It is missing the term "Default"

File: azure-pipelines.yml [hunk 6]
```diff
@@ -16,14 +16,54 @@ jobs:
...
+        colcon build --symlink-install --packages-up-to rmw_opendds_cpp
+    displayName:  build up to rmw_opendds_cpp
+  - script: |
+        cd $(Agent.BuildDirectory)
+        source /opt/OpenDDS/setenv.sh
+        source /opt/ros/eloquent/setup.bash
+        source install/local_setup.bash
+        printenv|grep DDS
+        printenv|grep ROS
+        colcon build --symlink-install --packages-up-to rcl_interfaces examples_rclcpp_minimal_publisher examples_rclcpp_minimal_subscriber
```
Comment (adamsj-ros) [inline azure-pipelines.yml]: Probably should add minimal service and client to get the srv files to build. I did manage to get the system_tests to build on a branch that would come after a clean build of the minimal examples of service and client.

File: azure-pipelines.yml [hunk 7]
```diff
@@ -16,14 +16,54 @@ jobs:
...
-        colcon build --symlink-install
-    displayName:  environment setup and build
+        export RMW_IMPLEMENTATION=rmw_opendds_cpp
+        env|grep RMW
+        ros2 run examples_rclcpp_minimal_subscriber subscriber_member_function_twenty 1> sub.out& 
+        process_id=$!
+        ros2 run examples_rclcpp_minimal_publisher publisher_member_function_twenty
+        wait $process_id
+        cat sub.out
+        grep "I heard" sub.out
```
Comment (adamsj-ros) [inline azure-pipelines.yml:68]: Probably not really part of this PR since we're just looking for clean build and not expecting a test to successfully run. We should be thinking about the next step for a runtime test for services if that means the minimal examples or getting the system tests to run. I'm not expecting a change for this PR at this time to run a test.

**Review by adamsj-ros — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml (hunk 7 above)
Comment (kuznetsovmoci) [reply azure-pipelines.yml:68]: I assumed this was not a PR specific change, but more a long term, general CI test setup. I am OK removing pub / sub tests though, if you think it does not belong to this package at this stage - can you please confirm?

**Review by mitza-oci — APPROVED** [review]

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 1 above)
Comment (adamsj-ros) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: It's preferable just to use the tools that ROS2 already offers. 

https://answers.ros.org/question/306194/ros1-and-ros2-linting/

Comment (mitza-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: Should that be in the CI scripts?

Comment (adamsj-ros) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: As a check, I guess so. I'd have to look into it more. But code beautifiers would be required to run prior to commit.

Comment (mitza-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: Right, I meant as a check -- just like we do in OpenDDS for some things

File: azure-pipelines.yml (hunk 7 above)
Comment (adamsj-ros) [reply azure-pipelines.yml:68]: I think pub/sub does belong here. It tests the other feature pub/sub that we want to preserve. Eventually we may want to add the client/service tests but only when it's closer to working with OpenDDS.

File: rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h (hunk 1 above)
Comment (lij-oci) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: Following the link Jeremy provided, I found some tools for checking the source code: cppcheck, and astyle. You can specify where the * or & go (left, middle, or right). However, I still can't find the ros style guidelines. If there is none on this particular point, we may as well use our OpenDDS guideline, at least consistent with ourselves.

Comment (adamsj-ros) [reply rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/service_type_support.h]: At this point for this PR, I think we should table this particular issue. This is the only remaining issue for this PR and I don't think it requires action to accept the PR. We definitely want a consistent style. There are tools to do so. We'll get there.

**Review by adamsj-ros — APPROVED** [review]
